<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Rarefying Phyloseq Data — rarefying • micromics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Rarefying Phyloseq Data — rarefying"><meta name="description" content="This function performs rarefaction on a normalized phyloseq object.
Rarefaction is based on the biomass of each sample to identify the minimum
sampling depth across the dataset. By rarefying all samples to this sampling
depth, we ensure that the data are normalized for sequencing effort. This
step prevents an overestimation of genus abundance in samples with higher
sequencing depths, allowing for more accurate and comparable results."><meta property="og:description" content="This function performs rarefaction on a normalized phyloseq object.
Rarefaction is based on the biomass of each sample to identify the minimum
sampling depth across the dataset. By rarefying all samples to this sampling
depth, we ensure that the data are normalized for sequencing effort. This
step prevents an overestimation of genus abundance in samples with higher
sequencing depths, allowing for more accurate and comparable results."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">micromics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/stijnteunissen/micromics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Rarefying Phyloseq Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/stijnteunissen/micromics/blob/master/R/rarefying.R" class="external-link"><code>R/rarefying.R</code></a></small>
      <div class="d-none name"><code>rarefying.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs rarefaction on a normalized <code>phyloseq</code> object.
Rarefaction is based on the biomass of each sample to identify the minimum
sampling depth across the dataset. By rarefying all samples to this sampling
depth, we ensure that the data are normalized for sequencing effort. This
step prevents an overestimation of genus abundance in samples with higher
sequencing depths, allowing for more accurate and comparable results.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rarefying</span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">physeq</span>, norm_method <span class="op">=</span> <span class="cn">NULL</span>, iteration <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-physeq">physeq<a class="anchor" aria-label="anchor" href="#arg-physeq"></a></dt>
<dd><p>A <code>phyloseq</code> object containing the microbiome data.
This is the input object that the function processes.</p></dd>


<dt id="arg-norm-method">norm_method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>A character string specifying the normalization method. Acceptable values are:</p><ul><li><p><code>NULL</code>: Use this option if no FCM or qPCR data is available, or if you wish to retain only relative abundances.</p></li>
<li><p><code>"fcm"</code>: Use this option if the data have been normalized using flow cytometry (FCM).</p></li>
<li><p><code>"qpcr"</code>: Use this option if the data have been normalized using quantitative PCR (qPCR).</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The rarefied <code>phyloseq</code> object is returned and also saved as an <code>.rds</code> file. The file name and location
depend on the specified normalization method:</p><ul><li><p>For <code>"fcm"</code>: <code>"project_name_phyloseq_asv_level_fcm_normalised_cell_concentration_rarefied.rds"</code></p></li>
<li><p>For <code>"qpcr"</code>: <code>"project_name_phyloseq_asv_level_qpcr_normalised_cell_concentration_rarefied.rds"</code></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<ul><li><p>For <code>"fcm"</code> normalization:</p><ul><li><p>Rarefies only the <code>fcm</code> normalized data, while the <code>copy number corrected</code> data remains unchanged.</p></li>
<li><p>Rarefies based on the calculated sampling depth, derived from the ratio of total reads per sample to the
estimated cell counts.</p></li>
<li><p>Uses a custom function (<code>avgrarefy</code>) to perform multiple iterations of rarefaction and averages the results.</p></li>
</ul></li>
<li><p>For <code>"qpcr"</code> normalization:</p><ul><li><p>Rarefies only the <code>qpcr</code> normalized data, while the <code>copy number corrected</code> data remains unchanged.</p></li>
<li><p>Rarefies based on the calculated sampling depth, derived from the ratio of total reads per sample to
predicted 16S rRNA gene copy numbers.</p></li>
<li><p>Uses the same <code>avgrarefy</code> function for averaging rarefied counts.</p></li>
</ul></li>
<li><p>The input values are subject to a scaling limit of 1e7. If input values exceed this limit due to the prior normalization,
all values are scaled down by a calculated scaling factor. The rarefied counts are rescaled back to their original scale
after rarefaction. However, due to this scaling and the rounding of scaled values, slight variations in the rarefied counts
may occur.</p></li>
</ul><p>The function uses parallel processing to improve the efficiency of rarefaction. It saves the rarefied <code>phyloseq</code>
object as an <code>.rds</code> file in the appropriate output folder.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>Ensure that the <code>phyloseq</code> object is properly normalized before applying this function.
Missing or invalid <code>rarefy_to</code> values will result in warnings and skipped samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Rarefy using FCM normalization</span></span></span>
<span class="r-in"><span><span class="va">rarefied_physeq</span> <span class="op">&lt;-</span> <span class="fu">rarefying</span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">normalised_physeq</span>, norm_method <span class="op">=</span> <span class="st">"fcm"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in log_message(paste("Step 9: rarefied data", paste(projects, collapse = ", ")),     log_file):</span> could not find function "log_message"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Rarefy using qPCR normalization</span></span></span>
<span class="r-in"><span><span class="va">rarefied_physeq</span> <span class="op">&lt;-</span> <span class="fu">rarefying</span><span class="op">(</span>physeq <span class="op">=</span> <span class="va">normalised_physeq</span>, norm_method <span class="op">=</span> <span class="st">"qpcr"</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in log_message(paste("Step 9: rarefied data", paste(projects, collapse = ", ")),     log_file):</span> could not find function "log_message"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stijn Teunissen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

